### PR TITLE
Point Cloud MixUp: linear interpolation data augmentation for OOD robustness

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

MixUp (Zhang et al., ICLR 2018) creates virtual training samples by linearly interpolating pairs of existing samples: x_mix = λ*x_1 + (1-λ)*x_2, y_mix = λ*y_1 + (1-λ)*y_2, where λ ~ Beta(α, α). This simple data augmentation technique has been shown to:
- Improve OOD generalization by 2-5% in classification and regression tasks
- Act as a vicinal risk minimizer — training on the manifold between data points
- Reduce overconfidence on out-of-distribution inputs

For CFD: interpolating between two airfoil configurations creates "virtual" intermediate geometries with plausible intermediate pressure fields. This directly expands the effective training distribution — a data augmentation approach per human directive #1860.

**Key constraint:** MixUp requires matching tensor shapes between paired samples. For TandemFoilSet, this means pairing samples with the SAME number of nodes (same mesh). If all samples share a common mesh topology, standard MixUp applies directly. If not, restrict to same-type pairs (tandem-tandem, single-single).

**Literature:**
- Zhang et al. "MixUp: Beyond Empirical Risk Minimization" (ICLR 2018)
- Verma et al. "Manifold Mixup: Better Representations by Interpolating Hidden States" (ICML 2019) — for intermediate layers
- Chen et al. "PointMixup: Augmentation for Point Clouds" (ECCV 2020) — MixUp for 3D point clouds

### Step 1: Implement MixUp in Training Loop

```python
def mixup_batch(x1, y1, x2, y2, alpha=0.2):
    """
    MixUp two samples with random interpolation factor.
    
    x1, x2: [N, D] input features (same shape required)
    y1, y2: [N, C] targets
    alpha: Beta distribution parameter (smaller = weaker mixing)
    
    Returns mixed inputs, targets, and lambda value.
    """
    lam = np.random.beta(alpha, alpha) if alpha > 0 else 1.0
    x_mix = lam * x1 + (1 - lam) * x2
    y_mix = lam * y1 + (1 - lam) * y2
    return x_mix, y_mix, lam
```

### Step 2: Integration with DataLoader

At each training step:
1. Draw a batch of samples normally
2. Create a shuffled copy of the batch (random permutation within the batch)
3. Apply MixUp between original and shuffled: `x_mix, y_mix, lam = mixup_batch(x_orig, y_orig, x_shuffled, y_shuffled, alpha=0.2)`
4. Train on the mixed batch

**Important:** MixUp should apply to:
- Input features (all 24 channels): x_mix = λ*x_1 + (1-λ)*x_2
- Targets (Ux, Uy, p): y_mix = λ*y_1 + (1-λ)*y_2
- Domain masks: mix by taking the union (if either sample has a surface node at position i, treat it as surface)
- PCGrad task assignment: assign mixed sample to the majority task of the two parents (e.g., if both are tandem → tandem; if mixed → single_foil)

**CRITICAL constraint:** Only mix samples with the same number of nodes. In the DataLoader, check that both samples in a pair have matching shapes. If they don't (e.g., single-foil vs tandem with different meshes), skip MixUp for that pair and use the original sample.

### Step 3: New Flags

- `--mixup` (bool, default False) — enable MixUp augmentation
- `--mixup_alpha 0.2` (float) — Beta distribution parameter. Smaller values = weaker mixing (closer to original samples). 0.2 is standard.
- `--mixup_prob 0.5` (float) — probability of applying MixUp to each batch. Not every batch needs mixing.

### Step 4: Training Runs

```
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/pointcloud-mixup-s42" \
  --wandb_group "pointcloud-mixup" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --mixup --mixup_alpha 0.2 --mixup_prob 0.5
```

Then seed 73 with same flags but `--seed 73`.

### Important Notes

- **asinh transform:** MixUp should apply AFTER the asinh pressure transform to targets (mix in transformed space). Mixing raw pressures then transforming is not equivalent to mixing transformed pressures.
- **Augmentation stacking:** MixUp applies ON TOP of existing augmentations (aoa_perturb, gap_stagger_sigma). The order is: (1) load sample, (2) apply standard augmentations, (3) apply MixUp to the augmented batch.
- **Validation:** Do NOT apply MixUp during validation — only training.
- **EMA:** EMA continues normally — it just sees the model trained on mixed samples.
- **Hard-node mining:** If using error-based hard-node mining, compute errors on the MIXED targets, not original targets. The mixed sample is the "ground truth" for that training step.
- **DSDF features:** When mixing input features, the DSDF gradient channels get interpolated between two geometries. This creates "virtual" geometries that don't correspond to physical meshes — but MixUp's power comes precisely from training on these impossible intermediate points.

## Baseline

Current best metrics (PR #2251, 2-seed avg):
- p_in: **11.891** (target: < 11.89)
- p_oodc: **7.561** (target: < 7.56)
- p_tan: **28.118** (target: < 28.12)
- p_re: **6.364** (target: < 6.36)
- W&B baseline runs: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Summary

Point Cloud MixUp (alpha=0.2, prob=0.5) **degraded all metrics** compared to baseline. This is a **negative result**.

### Metrics

| Metric | Baseline (2-seed avg) | S42 (MixUp) | S73 (MixUp) | 2-seed avg | Δ vs baseline |
|--------|----------------------|-------------|-------------|------------|---------------|
| p_in | **11.891** | 13.8 | 14.5 | 14.15 | **+19.0%** |
| p_oodc | **7.561** | 10.4 | 10.4 | 10.40 | **+37.5%** |
| p_tan | **28.118** | 29.8 | 30.2 | 30.00 | **+6.7%** |
| p_re | **6.364** | 7.5 | 7.6 | 7.55 | **+18.6%** |

### Run Details

**Seed 42** (W&B: `en8h216x`)
- Best epoch: 143 (of 143), val_loss = 0.4249
- mae_surf_p: in=13.8, tan=29.8, oodc=10.4, re=7.5
- Peak VRAM: 46.8 GB

**Seed 73** (W&B: `xl5pkdwv`)
- Best epoch: 143 (of 143), val_loss = 0.4321
- mae_surf_p: in=14.5, tan=30.2, oodc=10.4, re=7.6
- Peak VRAM: 46.8 GB

**Training:** 143 epochs in 180 min (wall clock limit)

**Command:**
```
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/pointcloud-mixup-s42" \
  --wandb_group "pointcloud-mixup" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --mixup --mixup_alpha 0.2 --mixup_prob 0.5
```

### What happened

MixUp consistently degraded all metrics across both seeds:
- **p_oodc worst hit** (+37.5%): OOD condition generalization significantly damaged
- **p_re second worst** (+18.6%): Reynolds number generalization hurt  
- **p_in degraded** (+19.0%): In-distribution performance dropped notably
- **p_tan mildest** (+6.7%): Tandem transfer least affected, but still worse
- Both seeds show consistent results (low variance) — this is a reliable negative signal, not noise

### Why it failed

1. **Domain mismatch with standard MixUp**: MixUp was designed for image classification where linear interpolation in pixel space creates plausible virtual examples. For CFD point clouds, interpolating between two geometries creates physically impossible configurations — the DSDF (signed distance field) gradients of the interpolated geometry don't correspond to any real airfoil surface. The model learns from these phantom geometries, which corrupts its understanding of real geometry-to-flow mappings.

2. **Already saturated augmentation**: The baseline already uses aoa_perturb (rotation augmentation), gap/stagger perturbation, and DSDF2 sigma augmentation. These are all physics-motivated augmentations that preserve geometric consistency. Adding MixUp on top introduces noise that competes with the meaningful augmentation signal.

3. **Target space mixing limitations**: Even though we mix normalized targets (after asinh transform) as requested, the mixed targets represent interpolated flow fields that don't satisfy the Navier-Stokes equations for the interpolated geometry. The model cannot learn consistent physics from these virtual examples.

4. **Surface mask dilution**: Taking the union of surface masks when mixing means the model treats more nodes as surface nodes, diluting the surface loss signal for actual surface predictions.

### Suggested follow-ups

1. **Manifold MixUp** (Verma et al., ICML 2019): Instead of mixing raw inputs, mix hidden representations at an intermediate layer of the Transolver. Interpolation in learned feature space may be more semantically meaningful than in input space.
2. **Same-type-only MixUp**: Only mix tandem-with-tandem and single-with-single to avoid cross-type geometric inconsistencies. Also restrict to similar AoA ranges.
3. **Consider this approach closed**: Standard input-space MixUp is fundamentally unsuited to physics-constrained point cloud regression where geometric consistency matters.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2282 because ADVISOR: Closing — Point Cloud MixUp severely degrades all metrics (p_in +19.0%, p_oodc +37.5%, p_tan +6.7%, p_re +18.6%).

**Root cause — standard MixUp is fundamentally unsuited to physics-constrained point cloud regression.** Linear interpolation between two CFD meshes creates physically impossible geometries where DSDF gradients don't correspond to any real airfoil surface. The model learns from phantom geometries that violate Navier-Stokes, corrupting its geometry-to-flow mapping. The existing augmentation suite (AoA perturb, gap/stagger, DSDF2 sigma) provides physics-consistent augmentation that MixUp noise competes against.

Excellent failure analysis — the surface mask dilution and target space mixing limitations are spot-on. Moving you to a fresh direction..
